sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -457,6 +470,17 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
     foreachBatch((batchDs: Dataset[T], batchId: Long) => function.call(batchDs, batchId))
   }
 
+  /**
+   * Specifies the underlying output table.
+   *
+   * @since 3.1.0
+   */
+  def table(tableName: String): DataStreamWriter[T] = {
```

Comment:
It's a bit weird to define `table` as a config method. I think it's better to follow `DataFrameWriter.saveAsTable` and make it an action. How about
```
def table(tableName: String): StreamingQuery = {
  this.source = SOURCE_NAME_TABLE
  this.tableName = tableName
  start()
}
```

Reply:
I have a bit different view on DataStreamWriter (and probably DataFrameWriter as well):

While we don't restrict the order, actually I think it's pretty much natural to have a flow, like `define a sink` -> `set options to the sink` -> `set options to the streaming query` -> `start the query`. (A couple of parts can be consolidated or the sequence can be swapped.)

```
df.writeStream
   .format("...")
   .option("...")
   .outputMode(...)
   .trigger(...)
   .start()
```

Now it looks to be simply arbitrary and something got mixed up. `checkpointLocation` isn't something being tied to the sink but we let end users to put into `option` which is also used for sink. `queryName` as well.

I intended the addition of `table` method as `defining a sink`, but if we'd like to care for tables specially, `DataFrameWriter.insertInto` would match the intention and I can change the method name to `insertInto` here as well. WDYT?

Reply:
DataFrameWriterV2 enforces the flow perfectly (let's put aside the flow branch for creating table), define a sink by providing table identifier, provide options, and decide which kind of write. The flow is uni-direction and no longer be arbitrary.

I'm feeling we should also have DataStreamWriterV2 to enforce the flow as well, but let's do this first and have more time to think about.

Reply:
I think we will have `DataStreamWriterV2` eventually (after we figure out how to design output mode). For now, it's more important to keep API consistency between batch and stream.

I don't have a strong opinion about the naming, `table` is fine. cc @xuanyuanking @zsxwing

Reply:
IMHO the exact match of the method from batch side looks to be `insertInto` (as it will handle the output mode and simply add/update the data instead of creating a table). Just naming it to `table` doesn't look to say it's an action - in DataFrameWriter it's `saveAsTable` instead of simply `table`.

Reply:
Ya. I also agree with the AS-IS config method `table` instead of action.

Reply:
then the API works differently between `DataFrameWriter` and `DataStreamWriter`, and users may be confusing.

I'm fine with `saveAsTable` if it's more like action.

Reply:
OK I'll change the name to saveAsTable and call start() there.